homeassistant/helpers/entity_component.py

```diff
@@ -88,6 +88,8 @@ def __init__(
 
         hass.data.setdefault(DATA_INSTANCES, {})[domain] = self
 
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self._async_shutdown)
```

Comment:
There are some cases where `setup`/`async_setup` is not called (eg. mailbox).

I think that `EVENT_HOMEASSISTANT_STOP` should be registered regardless.

Reply:
I think this is a side effect and we should avoid side effects in init methods.

Reply:
In that case, then maybe solving this in #90902 and #90917 is better after all.

Reply:
I think we should have a common method on the `EntityComponent` that sets up the listener. We'll need to do this for many entity integrations that won't support platform YAML config. Eg, I'm working on the stt integration right now that needs this.

Reply:
I couldn't decide if:
- it should be a new method (eg. `register_shutdown`)
- the current `async_setup` method should have an implicit check with `config` optional
- the current `async_setup` method should have an explicit check with a new argument (`supports_platform_yaml: bool = True`)

In the end I went with an implicit check as I feel it will be easier to maintain.